pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -137,15 +144,42 @@ func (binder *PersistentVolumeClaimBinder) updateClaim(oldObj, newObj interface{
 	binder.lock.Lock()
 	defer binder.lock.Unlock()
 	newClaim := newObj.(*api.PersistentVolumeClaim)
-	err := syncClaim(binder.volumeIndex, binder.client, newClaim)
+	err := syncClaim(binder.volumeIndex, binder.client, newClaim, binder.provisionedVolumes)
 	if err != nil {
 		glog.Errorf("PVClaimBinder could not update claim %s: %+v", newClaim.Name, err)
 	}
 }
 
-func syncVolume(volumeIndex *persistentVolumeOrderedIndex, binderClient binderClient, volume *api.PersistentVolume) (err error) {
+func (binder *PersistentVolumeClaimBinder) deleteClaim(obj interface{}) {
+	binder.lock.Lock()
+	defer binder.lock.Unlock()
+	if _, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		glog.V(5).Info("Missed delete event for a persistent volume claim")
+		return
+	}
+	claim, _ := obj.(*api.PersistentVolumeClaim)
+	if claim == nil {
+		glog.V(5).Infof("Removing claim %s from map of provisioned claims", ClaimToProvisionableKey(claim))
+		return
+	}
+	if _, exists := binder.provisionedVolumes[ClaimToProvisionableKey(claim)]; exists {
+		glog.V(5).Infof("Removing claim %s from map of provisioned claims", ClaimToProvisionableKey(claim))
+		delete(binder.provisionedVolumes, ClaimToProvisionableKey(claim))
+	}
+}
+
+func syncVolume(volumeIndex *persistentVolumeOrderedIndex, binderClient binderClient, volume *api.PersistentVolume, provisionedVolumes map[string]string) (err error) {
```

Comment:
I think it obscures the point of this method to have to inject so much of the binder's state.